src/client.rs

```diff
@@ -285,71 +280,181 @@ impl Client {
+                if !self.trust_local {
+                    debug!("Verify signature of local data.");
+                    self.verifier.verify(&collection)?;
+                }
 
+                Ok(collection.records)
+            }
+            None => {
+                if self.sync_if_empty {
+                    debug!("Synchronize data, without knowning which timestamp to expect.");
+                    let collection = self.sync(None)?;
+                    return Ok(collection.records);
+                }
+                // TODO: this empty list should be «qualified». Is it empty because never synced
+                // or empty on the server too. (see Normandy suitabilities).
+                debug!("Local data is empty or malformed.");
+                Ok(Vec::new())
+            }
+        }
+    }
+
+    /// Synchronize the local storage with the content of the server for this collection.
+    ///
+    /// # Behaviour
+    /// * If stored data is up-to-date and signature of local data valid, then return local content;
+    /// * Otherwise fetch content from server, merge with local content, verify signature, and return records;
+    ///
+    /// # Errors
+    /// If an error occurs while fetching or verifying records, a [`ClientError`] is returned.
+    pub fn sync<T>(&mut self, expected: T) -> Result<Collection, ClientError>
+    where
+        T: Into<Option<u64>>,
+    {
+        let storage_key = format!("{}/{}:collection", self.bucket_name, self.collection_name);
+
+        debug!("Retrieve from storage with key={:?}", storage_key);
+        let stored_bytes: Vec<u8> = self
+            .storage
+            .retrieve(&storage_key)
+            .unwrap_or(None)
+            .unwrap_or_else(Vec::new);
         let stored: Option<Collection> = serde_json::from_slice(&stored_bytes).unwrap_or(None);
 
-        // Note: when we implement the `sync()` method in #28,
-        // we will remove this, and `.get()` will return the current content of cache if it
-        // is not empty.
-        let remote_timestamp = get_latest_change_timestamp(
-            &self.server_url,
-            &self.bucket_name,
-            &self.collection_name,
-        )?;
+        let remote_timestamp = match expected.into() {
+            Some(v) => v,
+            None => {
+                debug!("Obtain current timestamp.");
+                get_latest_change_timestamp(
+                    &self.server_url,
+                    &self.bucket_name,
+                    &self.collection_name,
+                )?
+            }
+        };
 
-        if let Some(collection) = stored {
+        if let Some(ref collection) = stored {
             let up_to_date = collection.timestamp == remote_timestamp;
             if up_to_date && self.verifier.verify(&collection).is_ok() {
                 debug!("Local data is up-to-date and valid.");
-                return Ok(collection.records);
+                return Ok(collection.to_owned());
```

Comment:
This reallocates on copied the whole collection, which seems unnecessary. Maybe use
```suggestion
                return Ok(stored.unwrap());
```
instead? Unwrapping is fine here, since we already know that `stored` is `Some(_)`.